### ISSUE TITLE
How to Show More val_batch{i}_labels ?

### ISSUE THREAD

**Issue by cccccworld** [ultralytics/ultralytics#10182]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I plan to conduct a more detailed comparison of the predictive and actual validation results of my model during the validation phase using YOLOv8, and therefore need to generate more batch images after performing the validation tasks. To achieve this goal, I have explored possible solutions. The current recommendation I have received is to delete the number of batch images generated by modifying a specific section of the code:

```python
if self.args.plots and batch_i<3:
    self.plot_val_samples(batch, batch_i)
    self.plot_predictions(batch, preds, batch_i)
```

However, despite removing the `<3`condition and changing the batch sizes during validation, the final validation still only generates files numbered `val_batch{0-3}`. The change in batch size only affects the number of detections within each image.

![Snipaste_2024-04-20_11-39-55](https://github.com/ultralytics/ultralytics/assets/112063646/3253256b-ba35-4cb9-bd1c-3088fd60b69c)

It doesn't have any change to solve my question.

In fact, I also debug the code by running val seperately.

```python
from ultralytics.models.yolo.detect import DetectionValidator

if __name__ == '__main__':
    args = dict(model='yolov8n.pt', data='coco8.yaml')
    validator = DetectionValidator(args=args)
    validator()
```
I found that the code has ended after the `batch_i` loop from `0 to 3` despite removing the `<3`, I am not sure what the problem is, can not show all the validation data

So I'm wondering how to tweak the code to produce more contrast images, preferably with the entire verification set to output.

Thank you very much! I sincerely hope that you can help me with my problem.

**Comment by glenn-jocher** [comment]
Hello and thanks for reaching out with your question!

It looks like you're on the right track with modifying the code to generate more validation batch images. However, the number of images being processed and plotted during validation isn't solely influenced by the `if self.args.plots and batch_i<3:` condition. It's a bit more intricate since it involves the data loader's batch size and the total number of batches available during validation.

To generate plots for more (or all) validation samples, you've correctly identified one part of the solution by adjusting the condition within `self.plot_val_samples` and `self.plot_predictions`. Other than removing the check for `batch_i < 3`, ensure that your validation dataset has a sufficient number of batches to iterate over. If you wish to visualize every batch (which might be a lot! 😄), you'll need to confirm:

1. The total number of images in your validation set.
2. The `batch_size` you're using for validation.

To visualize all validation samples, you can try setting the `batch_size` to 1, which will process and potentially plot each image individually. Remember, however, that this might be computationally intensive and can result in a very large number of output images depending on the size of your validation set.

Here's a quick example to adjust batch size for the validation phase:

```python
if __name__ == '__main__':
    args = dict(model='yolov8n.pt', data='coco8.yaml', batch_size=1)  # Set batch_size to 1
    validator = DetectionValidator(args=args)
    validator()
```

Make sure you have enough storage space for all generated images and consider the computational cost before running the validation.

If the issue persists or if you encounter any errors, providing more context about the errors or behavioral inconsistencies will help pinpoint the issue further. 

I hope this addresses your concern. Looking forward to hearing how it goes! 🚀

**Comment by Love-syntacticSugar** [comment]
@cccccworld hello，Recently, I have also been studying the plotting issue and have encountered the same problem as yours. I believe this is due to the use of Python threading locks in the plot_images method (you will find a thread decorator inside this method). Since Python threads can handle at most three threads (I'm guessing), only three images can be plotted, even if you remove the condition batch_i < 3.

The solution to this problem is to add the parameter threaded=False to the plot_val_samples method in the detect package and add the parameter plot_predictions to the plot_predictions method in the obb package. In this way, all the code will run in a single thread (I have personally tried it and found it effective, and I asked the author in https://github.com/ultralytics/ultralytics/issues/13101, and there should be no problem, but the obvious drawback of this approach is slowness).
I hope it was helpful for you

**Comment by adriengoleb** [comment]
Hello,


I'm working with YOLOv8 and I'm trying to modify the plot_predictions function to save predictions per frame rather than per batch. However, I am getting a persistent error with ‘cv2.resize’ :


cv2.error: OpenCV(4.10.0) D:\a\opencv-python\opencv-python\opencv\modules\imgproc\src\resize.cpp:3789: error: (-215:Assertion failed) !dsize.empty() in function ‘cv::hal::resize’

This error occurs when I try to resize an image before saving it. I've already implemented several checks to ensure that the image dimensions (w_new and h_new) are valid, but the error persists :

```
def plot_images_without_mosaic(
    images: Union[torch.Tensor, np.ndarray],
    batch_idx: Union[torch.Tensor, np.ndarray],
    cls: Union[torch.Tensor, np.ndarray],
    bboxes: Union[torch.Tensor, np.ndarray] = np.zeros(0, dtype=np.float32),
    confs: Optional[Union[torch.Tensor, np.ndarray]] = None,
    masks: Union[torch.Tensor, np.ndarray] = np.zeros(0, dtype=np.uint8),
    kpts: Union[torch.Tensor, np.ndarray] = np.zeros((0, 51), dtype=np.float32),
    paths: Optional[List[str]] = None,
    fname: str = "images.jpg",
    names: Optional[Dict[int, str]] = None,
    on_plot: Optional[Callable] = None,
    max_size: int = 1920,
    max_subplots: int = 16,
    save: bool = True,
    conf_thres: float = 0.25,
) -> Optional[np.ndarray]:
    """
    Plot images with labels, bounding boxes, masks, and keypoints.

    Args:
        images: Batch of images to plot. Shape: (batch_size, channels, height, width).
        batch_idx: Batch indices for each detection. Shape: (num_detections,).
        cls: Class labels for each detection. Shape: (num_detections,).
        bboxes: Bounding boxes for each detection. Shape: (num_detections, 4) or (num_detections, 5) for rotated boxes.
        confs: Confidence scores for each detection. Shape: (num_detections,).
        masks: Instance segmentation masks. Shape: (num_detections, height, width) or (1, height, width).
        kpts: Keypoints for each detection. Shape: (num_detections, 51).
        paths: List of file paths for each image in the batch.
        fname: Output filename for the plotted image grid.
        names: Dictionary mapping class indices to class names.
        on_plot: Optional callback function to be called after saving the plot.
        max_size: Maximum size of the output image grid.
        max_subplots: Maximum number of subplots in the image grid.
        save: Whether to save the plotted image grid to a file.
        conf_thres: Confidence threshold for displaying detections.

    Returns:
        np.ndarray: Plotted image as a numpy array if save is False, None otherwise.
    """
    if isinstance(images, torch.Tensor):
        images = images.cpu().float().numpy()
    if isinstance(cls, torch.Tensor):
        cls = cls.cpu().numpy()
    if isinstance(bboxes, torch.Tensor):
        bboxes = bboxes.cpu().numpy()
    if isinstance(masks, torch.Tensor):
        masks = masks.cpu().numpy().astype(int)
    if isinstance(kpts, torch.Tensor):
        kpts = kpts.cpu().numpy()
    if isinstance(batch_idx, torch.Tensor):
        batch_idx = batch_idx.cpu().numpy()

    bs, _, h, w = images.shape  # batch size, _, height, width
    if np.max(images[0]) <= 1:
        images *= 255  # de-normalise (optional)

    # Process each image individually
    for i in range(bs):
        img = images[i].transpose(1, 2, 0).astype(np.uint8)  # Convert to HWC and uint8

        # Resize if necessary
        scale = max_size / max(h, w)

        global w_new, h_new
        with lock:

            if scale < 1:
                h_new, w_new = int(h * scale), int(w * scale)
                img = cv2.resize(img, (w_new, h_new))

        # Initialize annotator
        fs = int((h + w) * 0.01)  # font size
        annotator = Annotator(img, line_width=round(fs / 10), font_size=fs, pil=True, example=names)

        # Get detections for this image
        idx = batch_idx == i
        classes = cls[idx].astype("int")
        labels = confs is None

        # Plot bounding boxes
        if len(bboxes):
            boxes = bboxes[idx]
            conf = confs[idx] if confs is not None else None  # check for confidence presence (label vs pred)
            if len(boxes):
                if boxes[:, :4].max() <= 1.1:  # if normalized with tolerance 0.1
                    boxes[..., [0, 2]] *= w  # scale to pixels
                    boxes[..., [1, 3]] *= h
                elif scale < 1:  # absolute coords need scale if image scales
                    boxes[..., :4] *= scale
                is_obb = boxes.shape[-1] == 5  # xywhr
                boxes = xywhr2xyxyxyxy(boxes) if is_obb else xywh2xyxy(boxes)
                for j, box in enumerate(boxes.astype(np.int64).tolist()):
                    c = classes[j]
                    color = colors(c)
                    c = names.get(c, c) if names else c
                    if labels or conf[j] > conf_thres:
                        label = f"{c}" if labels else f"{c} {conf[j]:.1f}"
                        annotator.box_label(box, label, color=color, rotated=is_obb)

        # Plot keypoints
        if len(kpts):
            kpts_ = kpts[idx].copy()
            if len(kpts_):
                if kpts_[..., 0].max() <= 1.01 or kpts_[..., 1].max() <= 1.01:  # if normalized with tolerance .01
                    kpts_[..., 0] *= w  # scale to pixels
                    kpts_[..., 1] *= h
                elif scale < 1:  # absolute coords need scale if image scales
                    kpts_ *= scale
                for j in range(len(kpts_)):
                    if labels or conf[j] > conf_thres:
                        annotator.kpts(kpts_[j], conf_thres=conf_thres)

        # Plot masks
        if len(masks):
            if idx.shape[0] == masks.shape[0]:  # overlap_masks=False
                image_masks = masks[idx]
            else:  # overlap_masks=True
                image_masks = masks[[i]]  # (1, 640, 640)
                nl = idx.sum()
                index = np.arange(nl).reshape((nl, 1, 1)) + 1
                image_masks = np.repeat(image_masks, nl, axis=0)
                image_masks = np.where(image_masks == index, 1.0, 0.0)

            im = np.asarray(annotator.im).copy()
            for j in range(len(image_masks)):
                if labels or conf[j] > conf_thres:
                    color = colors(classes[j])
                    mh, mw = image_masks[j].shape
                    if mh != h or mw != w:
                        mask = image_masks[j].astype(np.uint8)
                        mask = cv2.resize(mask, (w, h))
                        mask = mask.astype(bool)
                    else:
                        mask = image_masks[j].astype(bool)
                    with contextlib.suppress(Exception):
                        im[mask] = im[mask] * 0.4 + np.array(color) * 0.6
            annotator.fromarray(im)

        # Save or return the annotated image
        if save:
            if paths:
                img_name = Path(paths[i]).stem
                output_path = str(Path(fname).parent / f"{img_name}_pred.jpg")
            else:
                output_path = fname
            annotator.im.save(output_path)
            if on_plot:
                on_plot(output_path)
        else:
            return np.asarray(annotator.im)
```

I want to modify the plot_predictions function to save each image with its predictions individually, rather than grouping the images together in a mosaic. **The error must come from the threading management.**

Any helps ? 

@Love-syntacticSugar Can you tell me what you mean by ‘add the parameter plot_predictions to the plot_predictions method in the obb package’? I have the same problem I think.

**Comment by deJQK** [comment]
Here is my code to solve this problem based on the description from @Love-syntacticSugar [here](https://github.com/ultralytics/ultralytics/issues/10182#issuecomment-2130664980):

```
from functools import partial

import ultralytics.utils.plotting
ultralytics.utils.plotting.plot_images = partial(ultralytics.utils.plotting.plot_images, threaded=False)

from ultralytics.engine.validator import BaseValidator
from ultralytics.models.yolo.detect.val import DetectionValidator

from ultralytics.data.utils import check_cls_dataset, check_det_dataset
from ultralytics.nn.autobackend import AutoBackend
from ultralytics.utils import LOGGER, TQDM, callbacks, colorstr, emojis
from ultralytics.utils.checks import check_imgsz
from ultralytics.utils.ops import Profile
from ultralytics.utils.torch_utils import de_parallel, select_device, smart_inference_mode

class MyBaseValidator(BaseValidator):
    @smart_inference_mode()
    def __call__(self, trainer=None, model=None):
        """
        Execute validation process, running inference on dataloader and computing performance metrics.

        Args:
            trainer (object, optional): Trainer object that contains the model to validate.
            model (nn.Module, optional): Model to validate if not using a trainer.

        Returns:
            stats (dict): Dictionary containing validation statistics.
        """
        self.training = trainer is not None
        augment = self.args.augment and (not self.training)
        if self.training:
            self.device = trainer.device
            self.data = trainer.data
            # Force FP16 val during training
            self.args.half = self.device.type != "cpu" and trainer.amp
            model = trainer.ema.ema or trainer.model
            model = model.half() if self.args.half else model.float()
            # self.model = model
            self.loss = torch.zeros_like(trainer.loss_items, device=trainer.device)
            self.args.plots &= trainer.stopper.possible_stop or (trainer.epoch == trainer.epochs - 1)
            model.eval()
        else:
            if str(self.args.model).endswith(".yaml") and model is None:
                LOGGER.warning("WARNING ⚠️ validating an untrained model YAML will result in 0 mAP.")
            callbacks.add_integration_callbacks(self)
            model = AutoBackend(
                weights=model or self.args.model,
                device=select_device(self.args.device, self.args.batch),
                dnn=self.args.dnn,
                data=self.args.data,
                fp16=self.args.half,
            )
            # self.model = model
            self.device = model.device  # update device
            self.args.half = model.fp16  # update half
            stride, pt, jit, engine = model.stride, model.pt, model.jit, model.engine
            imgsz = check_imgsz(self.args.imgsz, stride=stride)
            if engine:
                self.args.batch = model.batch_size
            elif not pt and not jit:
                self.args.batch = model.metadata.get("batch", 1)  # export.py models default to batch-size 1
                LOGGER.info(f"Setting batch={self.args.batch} input of shape ({self.args.batch}, 3, {imgsz}, {imgsz})")

            if str(self.args.data).split(".")[-1] in {"yaml", "yml"}:
                self.data = check_det_dataset(self.args.data)
            elif self.args.task == "classify":
                self.data = check_cls_dataset(self.args.data, split=self.args.split)
            else:
                raise FileNotFoundError(emojis(f"Dataset '{self.args.data}' for task={self.args.task} not found ❌"))

            if self.device.type in {"cpu", "mps"}:
                self.args.workers = 0  # faster CPU val as time dominated by inference, not dataloading
            if not pt:
                self.args.rect = False
            self.stride = model.stride  # used in get_dataloader() for padding
            self.dataloader = self.dataloader or self.get_dataloader(self.data.get(self.args.split), self.args.batch)

            model.eval()
            model.warmup(imgsz=(1 if pt else self.args.batch, 3, imgsz, imgsz))  # warmup

        self.run_callbacks("on_val_start")
        dt = (
            Profile(device=self.device),
            Profile(device=self.device),
            Profile(device=self.device),
            Profile(device=self.device),
        )
        bar = TQDM(self.dataloader, desc=self.get_desc(), total=len(self.dataloader))
        self.init_metrics(de_parallel(model))
        self.jdict = []  # empty before each val
        for batch_i, batch in enumerate(bar):
            self.run_callbacks("on_val_batch_start")
            self.batch_i = batch_i
            # Preprocess
            with dt[0]:
                batch = self.preprocess(batch)

            # Inference
            with dt[1]:
                preds = model(batch["img"], augment=augment)

            # Loss
            with dt[2]:
                if self.training:
                    self.loss += model.loss(batch, preds)[1]

            # Postprocess
            with dt[3]:
                preds = self.postprocess(preds)

            self.update_metrics(preds, batch)
            if self.args.plots:
                self.plot_val_samples(batch, batch_i)
                self.plot_predictions(batch, preds, batch_i)

            self.run_callbacks("on_val_batch_end")
        stats = self.get_stats()
        self.check_stats(stats)
        self.speed = dict(zip(self.speed.keys(), (x.t / len(self.dataloader.dataset) * 1e3 for x in dt)))
        self.finalize_metrics()
        self.print_results()
        self.run_callbacks("on_val_end")
        if self.training:
            model.float()
            results = {**stats, **trainer.label_loss_items(self.loss.cpu() / len(self.dataloader), prefix="val")}
            return {k: round(float(v), 5) for k, v in results.items()}  # return results as 5 decimal place floats
        else:
            LOGGER.info(
                "Speed: {:.1f}ms preprocess, {:.1f}ms inference, {:.1f}ms loss, {:.1f}ms postprocess per image".format(
                    *tuple(self.speed.values())
                )
            )
            if self.args.save_json and self.jdict:
                with open(str(self.save_dir / "predictions.json"), "w", encoding="utf-8") as f:
                    LOGGER.info(f"Saving {f.name}...")
                    json.dump(self.jdict, f)  # flatten and save
                stats = self.eval_json(stats)  # update stats
            if self.args.plots or self.args.save_json:
                LOGGER.info(f"Results saved to {colorstr('bold', self.save_dir)}")
            return stats

class MyDetectionValidator(DetectionValidator, MyBaseValidator):
    pass

...

def main():
    ...
    metrics = model.val(
        validator=MyDetectionValidator,
        ...
    )
```
